### PR TITLE
Adding TorchMD-NET

### PR DESCRIPTION
**Pull request by RaulPPelaez**
Checklist
- [x] Title of this PR is meaningful: e.g. "Adding my_nifty_package", not "updated meta.yaml".
- [x] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/5eddbd7fc9d1502169089da06c3688d9759be978/recipes/example/meta.yaml#L64-L73) for an example).
- [x] Source is from official source.
- [x] Package does not vendor other packages. (If a package uses the source of another package, they should be separate packages or the licenses of all packages need to be packaged).
- [x] If static libraries are linked in, the license of the static library is packaged.
- [x] Package does not ship static libraries. If static libraries are needed, [follow CFEP-18](https://github.com/conda-forge/cfep/blob/main/cfep-18.md).
- [x] Build number is 0.
- [x] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your recipe (see [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos) for more details).
- [x] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there.
- [x] When in trouble, please check our [knowledge base documentation](https://conda-forge.org/docs/maintainer/knowledge_base.html) before pinging a team.

### CODE REVIEW

File: recipes/torchmd-net/meta.yaml
Comment (raimis) [inline recipes/torchmd-net/meta.yaml:65]: Include me and @giadefa

**Comment by RaulPPelaez** [comment]
cc @raimis, @giadefa, you are listed as maintainers of this recipe and you must explicitly agree to be one by commenting here.

**Comment by raimis** [comment]
I agree to be a maintainer of this package.

**Comment by RaulPPelaez** [comment]
I am compiling a pytorch C++ extension in setup.py, compilation succeeds but after installation this error pops up:

```shell
  ERROR (torchmd-net,lib/python3.9/site-packages/torchmdnet/neighbors/torchmdnet_neighbors.so): $RPATH/libc10.so not found in packages, sysroot(s) nor the missing_dso_whitelist.
.. is this binary repackaging?
  ERROR (torchmd-net,lib/python3.9/site-packages/torchmdnet/neighbors/torchmdnet_neighbors.so): $RPATH/libtorch_cpu.so not found in packages, sysroot(s) nor the missing_dso_whitelist.
.. is this binary repackaging?
 ```
Pytorch is listed as a host, build and run dependency:
```yaml
  build:
    - python
    - {{ compiler('c') }}
    - {{ compiler('cxx') }}
    - {{ compiler('cuda') }}  # [cuda_compiler_version not in (undefined, 'None')]
    - pytorch
    - pytorch =*={{ torch_proc_type }}*
  host:
    - python
    - pip
    - pytorch
    - pytorch =*={{ torch_proc_type }}*
  run:
    - python
    - h5py
    - matplotlib-base
    - nnpops
    - pytorch
    - pytorch =*={{ torch_proc_type }}*
    - pytorch_cluster
    - pytorch_geometric
    - pytorch_scatter
    - pytorch_sparse
    - lightning
    - pydantic <2
    - torchmetrics
    - tqdm
  run_constrained:
    # additional run constraint to the one from the (version-only) run_export;
    # constraining the CPU builds to CPU pytorch isn't 100% necessary, but cleaner
    - pytorch =*={{ torch_proc_type }}*
```

I do not know what the problem is -.- 
Any ideas?

**Comment by raimis** [comment]
You need just some linker magic (https://github.com/conda-forge/openmm-torch-feedstock/blob/d97b9105196cfd381e4252a19a4ea20bc38b149c/recipe/meta.yaml#L23):
```yaml
  rpaths:
    - lib/
    - {{ SP_DIR }}/torch/lib
```

**Comment by RaulPPelaez** [comment]
The linux_64 recipe fails abruptly:
```shell
tests/test_wrappers.py::test_atom_filter[tensornet-2] PASSED
/home/conda/staged-recipes/build_artifacts/torchmd-net_1695223680319/test_tmp/run_test.sh: line 7:  3452 Killed                  pytest -v -x -s tests
Tests failed for torchmd-net-0.3.0-py39h42bb6ba_0.conda - moving package to /home/conda/staged-recipes/build_artifacts/broken
TESTS FAILED: torchmd-net-0.3.0-py39h42bb6ba_0.conda
##[error]Bash exited with code '1'.
Finishing: Run docker build
```
However, no test fails.

File: recipes/torchmd-net/meta.yaml
Comment (raimis) [inline recipes/torchmd-net/meta.yaml:74]: Summary is still from OpenMM-Torch

**Comment by RaulPPelaez** [comment]
All checks are passing. Lets start the review process

**Comment by RaulPPelaez** [comment]
@conda-forge/python, ready for review!

**Comment by RaulPPelaez** [comment]
@conda-forge-admin, please ping team

**Comment by ocefpaf** [comment]
> @jaimergp @mikemhenry any idea how long it takes to get the revision of a new package?

It depends on the complexity of the recipe and the availability of our volunteers reviewers who don't get paid to do this BTW.

This recipe has compiled code and uses `cuda`, it is not an easy one. Maybe @jakirkham can review or ask someone with cuda experience to review it for you.

File: recipes/torchmd-net/meta.yaml
Comment (mikemhenry) [reply recipes/torchmd-net/meta.yaml:57]: Yes those should be consolidated (not sure if it will merge them).

**Comment by mikemhenry** [comment]
Yes what @ocefpaf said, everything is volunteer and the more complex the longer it takes. @jakirkham would be great to review this. In the meantime what you can do as a stop gap is go here: https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=785141&view=artifacts&pathAsName=false&type=publishedArtifacts download the build artifacts, then upload them to your own channel on anaconda, then users can still download the packages/use/test them while we wait on the review. I can help with this if you want. Then users will be able to do something like `micromamba install -c your-channel -c conda-forge torchmd-net`

**Comment by RaulPPelaez** [comment]
Cool! Did not know that was an option.
Thanks for the good work guys. We are aware CUDA+torch is never an easy feat to review :P

**Comment by RaulPPelaez** [comment]
In order to overcome this pytorch issue https://github.com/pytorch/pytorch/issues/109719 we need to compile an inline extension during runtime, which requires gxx as a runtime dependency. 
The idea of inline extensions is appealing, but I am not sure it is worth it here... Probably best to precompile along the neighbors one.

**Comment by mikemhenry** [comment]
@jakirkham I was wondering if you have time to take a look at this recipe and could give it a review? Thanks!

**Comment by RaulPPelaez** [comment]
Failing build is due to an overprotective test that sometimes fails... The failure can be disregarded.

**Comment by raimis** [comment]
The package for 0.7.1 are uploaded: https://anaconda.org/acellera/torchmd-net/files?version=0.7.1

File: recipes/torchmd-net/meta.yaml
Comment (xhochy) [inline recipes/torchmd-net/meta.yaml:29]:
```suggestion
```

File: recipes/torchmd-net/meta.yaml
Comment (xhochy) [inline recipes/torchmd-net/meta.yaml:30]:
```suggestion
```

File: recipes/torchmd-net/meta.yaml
Comment (xhochy) [inline recipes/torchmd-net/meta.yaml:25]: Probably only the compilers should be in build
```suggestion
```

File: recipes/torchmd-net/meta.yaml
Comment (xhochy) [inline recipes/torchmd-net/meta.yaml:20]: Is this really needed? If so: Can you post the error? Mostly it works without the rpaths spec.

**Review by xhochy — CHANGES_REQUESTED** [review]

File: recipes/torchmd-net/meta.yaml
Comment (RaulPPelaez) [reply recipes/torchmd-net/meta.yaml:20]: Thank you very much for your review @xhochy !
The error is this here:
https://github.com/conda-forge/staged-recipes/pull/24021#issuecomment-1727782866
I will try again and see if the error has been solved since then.

Comment (RaulPPelaez) [reply recipes/torchmd-net/meta.yaml:20]: I tried to remove the lines and the old error is reproduced, the package compiles fine, but then this is printed:
```shell


Packaging torchmd-net
Packaging torchmd-net-0.8.0-py39h42bb6ba_0
compiling .pyc files...
number of files: 87
Warning: rpath /home/conda/staged-recipes/build_artifacts/torchmd-net_1698057723522/_build_env/lib is outside prefix /home/conda/staged-recipes/build_artifacts/torchmd-net_1698057723522/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold (removing it)
   INFO: sysroot: '/home/conda/staged-recipes/build_artifacts/torchmd-net_1698057723522/_build_env/x86_64-conda-linux-gnu/sysroot/' files: '['/home/conda/staged-recipes/build_artifacts/torchmd-net_1698057723522/_build_env/x86_64-conda-linux-gnu/sysroot/usr/share/zoneinfo/zone.tab', '/home/conda/staged-recipes/build_artifacts/torchmd-net_1698057723522/_build_env/x86_64-conda-linux-gnu/sysroot/usr/share/zoneinfo/tzdata.zi', '/home/conda/staged-recipes/build_artifacts/torchmd-net_1698057723522/_build_env/x86_64-conda-linux-gnu/sysroot/usr/share/zoneinfo/right/Zulu', '/home/conda/staged-recipes/build_artifacts/torchmd-net_1698057723522/_build_env/x86_64-conda-linux-gnu/sysroot/usr/share/zoneinfo/right/WET']'
  ERROR (torchmd-net,lib/python3.9/site-packages/torchmdnet/extensions/torchmdnet_extensions.so): $RPATH/libc10.so not found in packages, sysroot(s) nor the missing_dso_whitelist.
.. is this binary repackaging?
  ERROR (torchmd-net,lib/python3.9/site-packages/torchmdnet/extensions/torchmdnet_extensions.so): $RPATH/libtorch_cpu.so not found in packages, sysroot(s) nor the missing_dso_whitelist.
.. is this binary repackaging?
   INFO (torchmd-net,lib/python3.9/site-packages/torchmdnet/extensions/torchmdnet_extensions.so): Needed DSO lib/libstdc++.so.6 found in conda-forge::libstdcxx-ng-13.2.0-h7e041cc_2
   INFO (torchmd-net,lib/python3.9/site-packages/torchmdnet/extensions/torchmdnet_extensions.so): Needed DSO lib/libgcc_s.so.1 found in conda-forge::libgcc-ng-13.2.0-h807b86a_2
   INFO (torchmd-net,lib/python3.9/site-packages/torchmdnet/extensions/torchmdnet_extensions.so): Needed DSO x86_64-conda-linux-gnu/sysroot/lib64/libc.so.6 found in CDT/compiler package conda-forge::sysroot_linux-64-2.12-he073ed8_16
WARNING (torchmd-net): interpreter (Python) package conda-forge::python-3.9.18-h0755675_0_cpython in requirements/run but it is not used (i.e. it is overdepending or perhaps statically linked? If that is what you want then add it to `build/ignore_run_exports`)
WARNING (torchmd-net): run-exports library package conda-forge::pytorch-2.0.0-cpu_mkl_py39h1d6e76c_103 in requirements/run but it is not used (i.e. it is overdepending or perhaps statically linked? If that is what you want then add it to `build/ignore_run_exports`)
Traceback (most recent call last):
  File "/home/conda/staged-recipes-copy/.ci_support/build_all.py", line 228, in <module>
    build_all(os.path.join(root_dir, "recipes"), args.arch)
  File "/home/conda/staged-recipes-copy/.ci_support/build_all.py", line 118, in build_all
    build_folders(recipes_dir, folders, arch, channel_urls)
  File "/home/conda/staged-recipes-copy/.ci_support/build_all.py", line 174, in build_folders
    conda_build.api.build([recipe], config=get_config(arch, channel_urls))
  File "/opt/conda/lib/python3.10/site-packages/conda_build/api.py", line 253, in build
    return build_tree(
  File "/opt/conda/lib/python3.10/site-packages/conda_build/build.py", line 3804, in build_tree
    packages_from_this = build(
  File "/opt/conda/lib/python3.10/site-packages/conda_build/build.py", line 2880, in build
    newly_built_packages = bundlers[pkg_type](output_d, m, env, stats)
  File "/opt/conda/lib/python3.10/site-packages/conda_build/build.py", line 2006, in bundle_conda
    files = post_process_files(metadata, initial_files)
  File "/opt/conda/lib/python3.10/site-packages/conda_build/build.py", line 1817, in post_process_files
    post_build(m, new_files, build_python=python)
  File "/opt/conda/lib/python3.10/site-packages/conda_build/post.py", line 1721, in post_build
    check_overlinking(m, files, host_prefix)
  File "/opt/conda/lib/python3.10/site-packages/conda_build/post.py", line 1605, in check_overlinking
    return check_overlinking_impl(
  File "/opt/conda/lib/python3.10/site-packages/conda_build/post.py", line 1576, in check_overlinking_impl
    raise OverLinkingError(overlinking_errors)
conda_build.exceptions.OverLinkingError: overlinking check failed 
['  ERROR (torchmd-net,lib/python3.9/site-packages/torchmdnet/extensions/torchmdnet_extensions.so): $RPATH/libc10.so not found in packages, sysroot(s) nor the missing_dso_whitelist.\n.. is this binary repackaging?', '  ERROR (torchmd-net,lib/python3.9/site-packages/torchmdnet/extensions/torchmdnet_extensions.so): $RPATH/libtorch_cpu.so not found in packages, sysroot(s) nor the missing_dso_whitelist.\n.. is this binary repackaging?']
##[error]Bash exited with code '1'.
Finishing: Run docker build
```

Comment (raimis) [reply recipes/torchmd-net/meta.yaml:20]: `pytorch` has its libraries not in the standard place.

**Comment by RaulPPelaez** [comment]
Now that NNPOps has an OSX build I gave the torchmd-net build a try, but it fails because it does not find the torch libraries:
```shell

   INFO: sysroot: '/Applications/Xcode_13.2.1.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.9.sdk/' files: '['usr/share/man/mann/zero.n', 'usr/share/man/mann/yencode.n', 'usr/share/man/mann/yaml.n', 'usr/share/man/mann/xsxp.n']'
   INFO (torchmd-net,lib/python3.11/site-packages/torchmdnet/extensions/torchmdnet_extensions.so): Needed DSO lib/python3.11/site-packages/torch/lib/libc10.dylib found in conda-forge::pytorch-2.0.0-cpu_mkl_py311h20fc508_103
   INFO (torchmd-net,lib/python3.11/site-packages/torchmdnet/extensions/torchmdnet_extensions.so): Needed DSO lib/python3.11/site-packages/torch/lib/libtorch_cpu.dylib found in conda-forge::pytorch-2.0.0-cpu_mkl_py311h20fc508_103
  ERROR (torchmd-net,lib/python3.11/site-packages/torchmdnet/extensions/torchmdnet_extensions.so): Needed DSO lib/libc++.1.dylib found in ['libcxx']
  ERROR (torchmd-net,lib/python3.11/site-packages/torchmdnet/extensions/torchmdnet_extensions.so): .. but ['libcxx'] not in reqs/run, (i.e. it is overlinking) (likely) or a missing dependency (less likely)
   INFO (torchmd-net,lib/python3.11/site-packages/torchmdnet/extensions/torchmdnet_extensions.so): Needed DSO /Applications/Xcode_13.2.1.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.9.sdk/usr/lib/libSystem.B.dylib found in $SYSROOT
WARNING (torchmd-net): interpreter (Python) package conda-forge::python-3.11.6-h30d4d87_0_cpython in requirements/run but it is not used (i.e. it is overdepending or perhaps statically linked? If that is what you want then add it to `build/ignore_run_exports`)
Traceback (most recent call last):
  File "/Users/runner/work/1/s/.ci_support/build_all.py", line 228, in <module>
    build_all(os.path.join(root_dir, "recipes"), args.arch)
  File "/Users/runner/work/1/s/.ci_support/build_all.py", line 118, in build_all
    build_folders(recipes_dir, folders, arch, channel_urls)
  File "/Users/runner/work/1/s/.ci_support/build_all.py", line 174, in build_folders
    conda_build.api.build([recipe], config=get_config(arch, channel_urls))
  File "/Users/runner/Miniforge3/lib/python3.10/site-packages/conda_build/api.py", line 253, in build
    return build_tree(
  File "/Users/runner/Miniforge3/lib/python3.10/site-packages/conda_build/build.py", line 3804, in build_tree
    packages_from_this = build(
  File "/Users/runner/Miniforge3/lib/python3.10/site-packages/conda_build/build.py", line 2880, in build
    newly_built_packages = bundlers[pkg_type](output_d, m, env, stats)
  File "/Users/runner/Miniforge3/lib/python3.10/site-packages/conda_build/build.py", line 2006, in bundle_conda
    files = post_process_files(metadata, initial_files)
  File "/Users/runner/Miniforge3/lib/python3.10/site-packages/conda_build/build.py", line 1817, in post_process_files
    post_build(m, new_files, build_python=python)
  File "/Users/runner/Miniforge3/lib/python3.10/site-packages/conda_build/post.py", line 1721, in post_build
    check_overlinking(m, files, host_prefix)
  File "/Users/runner/Miniforge3/lib/python3.10/site-packages/conda_build/post.py", line 1605, in check_overlinking
    return check_overlinking_impl(
  File "/Users/runner/Miniforge3/lib/python3.10/site-packages/conda_build/post.py", line 1576, in check_overlinking_impl
    raise OverLinkingError(overlinking_errors)
conda_build.exceptions.OverLinkingError: overlinking check failed 
["  ERROR (torchmd-net,lib/python3.11/site-packages/torchmdnet/extensions/torchmdnet_extensions.so): .. but ['libcxx'] not in reqs/run, (i.e. it is overlinking) (likely) or a missing dependency (less likely)"]
##[error]Bash exited with code '1'.
Finishing: Run OSX build
```
I will turn it off for now.

**Comment by carterbox** [comment]
> 
> I will turn it off for now.

This error is telling you that `torchmdnet_extensions.so` is either overlinking to `libcxx` in which case you need to fix your build process so that you don't link to unneeded libraries. OR that it does need libcxx and you just need to add the `libcxx # [osx]` to the run requirements section.

It doesn't say it can't find the cxx libraries. It tells you that they are available, but that they are not listed in the requirements section of your recipe.

**Comment by isuruf** [comment]
This error was a packaging bug and is fixed now.

**Pull request merged**
